client/state-db/src/pruning.rs

```diff
@@ -47,153 +51,440 @@ pub struct RefWindow<BlockHash: Hash, Key: Hash> {
 	/// Number of calls of `prune_one` after
 	/// last call `apply_pending` or `revert_pending`
 	pending_prunings: usize,
-	/// Keep track of re-inserted keys and do not delete them when pruning.
-	/// Setting this to false requires backend that supports reference
-	/// counting.
-	count_insertions: bool,
 }
 
-#[derive(Debug, PartialEq, Eq, parity_util_mem_derive::MallocSizeOf)]
-struct DeathRow<BlockHash: Hash, Key: Hash> {
-	hash: BlockHash,
-	journal_key: Vec<u8>,
-	deleted: HashSet<Key>,
-}
-
-#[derive(Encode, Decode)]
-struct JournalRecord<BlockHash: Hash, Key: Hash> {
-	hash: BlockHash,
-	inserted: Vec<Key>,
-	deleted: Vec<Key>,
-}
-
-fn to_journal_key(block: u64) -> Vec<u8> {
-	to_meta_key(PRUNING_JOURNAL, &block)
+/// `DeathRowQueue` used to keep track of blocks in the pruning window, there are two flavors:
+/// - `Mem`, used when the backend database do not supports reference counting, keep all
+/// 	blocks in memory, and keep track of re-inserted keys to not delete them when pruning
+/// - `DbBacked`, used when the backend database supports reference counting, only keep
+/// 	a few number of blocks in memory and load more blocks on demand
+#[derive(parity_util_mem_derive::MallocSizeOf)]
+enum DeathRowQueue<BlockHash: Hash, Key: Hash, D: MetaDb> {
+	Mem {
+		/// A queue of keys that should be deleted for each block in the pruning window.
+		death_rows: VecDeque<DeathRow<BlockHash, Key>>,
+		/// An index that maps each key from `death_rows` to block number.
+		death_index: HashMap<Key, u64>,
+	},
+	DbBacked {
+		// The backend database
+		db: D,
```

Comment:
Passing a `&Db` param around may need to changing lot of interfaces even if `&Db` is not needed at all (like `DeathRowQueue::Mem`).

How about we remove this Clone bound and pass a `D` instead of `&D`, such that we don't need to do the cloning in `RefWindow::new` but in `client/db/src/lib.rs` where we can see it is wrapped by an `Arc`.

Reply:
Yes Arc is what I thought at first, but it will be a bit of a rework around the tests, but thinking twice having Arc<RwLock<TestDb is no issue.